common/api/core-geometry.api.md

```diff
@@ -5919,15 +5919,16 @@ export class Vector2d extends XY implements BeJSONFunctions {
     fractionOfProjectionToVector(target: Vector2d, defaultFraction?: number): number;
     static fromJSON(json?: XYProps): Vector2d;
     interpolate(fraction: number, right: Vector2d, result?: Vector2d): Vector2d;
-    isParallelTo(other: Vector2d, oppositeIsParallel?: boolean): boolean;
-    isPerpendicularTo(other: Vector2d): boolean;
+    isParallelTo(other: Vector2d, oppositeIsParallel?: boolean, radianSquaredTol?: number, distanceSquaredTol?: number): boolean;
+    isPerpendicularTo(other: Vector2d, returnValueIfAnInputIsZeroLength?: boolean, radianSquaredTol?: number, distanceSquaredTol?: number): boolean;
```

Comment:
Please consider preferring to take an `options` object instead of adding multiple new additional optional arguments.
Then callers do not have to explicitly specify `undefined` for options they don't care about, and do not have to be careful to match order of arguments.
```
interface IsPerpendicularToArgs {
  returnValueIfAnInputIsZeroLength?: boolean;
  radianSquaredTol?: number;
  distancedScquredTol?: number;
}

class Vector2d {
  public isPerpendicularTo(other: Vector3d, options: IsPerpendicularArgs): boolean {
    // Proceed.
  }
}

Reply:
I can use `options` but there are 4 similar functions that I have changed in this PR (2d and 3d versions of `isPerpendicularTo` and `isParallelTo`) and I want to keep them consistent. I dont like one of them uses `options` and 3 others dont. In fact I have added 3 parameters to the 2d version of `isPerpendicularTo` to make it consistent with the 3d version.

I think you want this change to reduce number of function arguments. Well, I can pack the 2 new parameters (`radianSquaredTol` and `distanceSquaredTol`) to an `options` object to reduce parameters by one. However, I cannot add more parameters to the `options` because that would lead to API change and would break users code.

Reply:
I saw the other functions. I was suggesting you change all 4.
I'm not sure what you mean about adding more parameters to `options` breaking the API - one primary motivator for using `options` is you can add new parameters to it in the future *without* breaking the API nor making it messier.

Reply:
@pmconne 
I can pack all 4 parameters into an `options` object if you tell me how to resolve the API change issue

Reply:
I believe they are referring to moving the existing parameters from the current public Vector3d methods into options when talking about breaking the api.

  public isParallelTo(other: Vector3d, oppositeIsParallel: boolean = false, returnValueIfAnInputIsZeroLength: boolean = false): boolean {
  public isPerpendicularTo(other: Vector3d, returnValueIfAnInputIsZeroLength: boolean = false): boolean {

While I agree that options are a good way to handle this and should be used for new apis, I think in this case the cat is already out of the bag and consistency between the 2d/3d methods is better. I also can't see more optonal args getting added, these now pretty much cover every case...

Reply:
I agree with @bbastings (and I think with you, @saeeedtorabi) - make each function take `returnValueIfAnInputIsZeroLength?: boolean, options?: { radianSquaredTol?: number, distanceSquaredTol?: number }`.

Reply:
If we start using `options` now, a user's code that has `isParallelTo(vector, true, true)` would not be compiled until it's changed to `isParallelTo(vector, argsOptions) `

However, your point is valid. Using parameters gives us the flexiility to add more and more parameters without API change and without making the arg list too long.

Reply:
Is this alternative too awkward: widen the type of the first optional arg to include the new options interface. This allows both signatures, with a bit of narrowing logic in the implementation, but no changes to existing callers.  E.g., `public isParallelTo(other: Vector3d, oppositeIsParallel: boolean | PerpParallelOptions = false, returnValueIfAnInputIsZeroLength: boolean = false): boolean` then inside the method, ignore the old API if the options are given in the first optional arg. Callers can then do isPerpendicularTo(other, opts).  All four options (existing and new) are in the opts interface.

Reply:
> isParallelTo(other: Vector3d, oppositeIsParallel: boolean | PerpParallelOptions = false,

I don't like this, it means ignoring the 2nd arg "returnValueIfAnInputIsZeroLength" when the 1st arg is an option. Just adding the existing args from the 3d version to 2d and adding the 2 new tolerances in options as Paul suggested seems cleanest at this point.

Reply:
I agree with @pmconne 
making those kind of API change is confusing

Reply:
can we add 2 new functions with the same names but 2 arguments (vector and `options`) and deprecate the old functions?

Reply:
The way function overloading works in TypeScript, no. You could add the new overloads, but you can't deprecate specific overloads. Again, ends up being more confusing for users than it's worth.

Reply:
Paul, you suggested adding
```typescript
options?: { radianSquaredTol?: number, distanceSquaredTol?: number }
```
​but we have default values for those so that's in fact
```typescript
options: { radianSquaredTol: number = Geometry.smallAngleRadiansSquared, distanceSquaredTol: number = Geometry.smallMetricDistanceSquared }
```
​but I dont see the need for options here. That only makes the function arguments longer and the code less readable compare to what it is right now.

Reply:
If do it the interface way as in @pmconne 's first suggestion, to minimally affect the arglist: `foo(a, b, c, options?: PerpParallelOptions)`.

Reply:
Interface fields cannot have default values.
You declare them as optional, and document the defaults.

Reply:
I laugh, but I also learn. Thanks @pmconne.